apps/dashr-daq-iv-tracer/app.R

```diff
@@ -0,0 +1,1508 @@
+  # '''
+  function(nClick, knobVal, measTriggered, storeNclick) {
+
+    if (is.list(nClick)) {
+      nClick <- 0
+    }
+    if (is.list(storeNclick)) {
+      storeNclick <- 0
+    }
+
+    if (nClick != storeNclick) {
+      return(TRUE)
+    } else {
+      return(FALSE)
+    }
+  }
+)
+
+
+app$callback(
+  output = list(id = "IV_graph", property = "figure"),
+  params = list(
+    input(id = "measure-display", property = "value"),
+    input(id = "clear-graph_ind", property = "value"),
+    state(id = "toggleTheme", property = "value"),
+    state(id = "measure-triggered", property = "value"),
+    state(id = "IV_graph", property = "figure"),
+    state(id = "source-choice-toggle", property = "value"),
+    state(id = "mode-choice-toggle", property = "value"),
+    state(id = "sweep-status", property = "value"),
+    state(id = "store-sourced-values", property = "data"),
+    state(id = "store-measured-values", property = "data")
+  ),
+
+  # ''' Update the IV-Graph '''
+  function(measuredVal,
+           clearGraph, # to fire callback on reset
+           theme,
+           measTriggered,
+           graphData,
+           srcChoice,
+           modeChoice,
+           swpOn,
+           sourceStore,
+           measStore) {
+
+    if (theme) {
+      theme <- "dark"
+    } else {
+      theme <- "light"
+    }
+
+    if (srcChoice) {
+      srcType <- "Current"
+    } else {
+      srcType <- "Voltage"
+    }
+
+    # Labels for sourced and measured quantities
+    sourceLabel <- GetSourceLabels(srcType)[1]
+    measureLabel <- GetSourceLabels(srcType)[2]
+    sourceUnit <- GetSourceUnits(srcType)[1]
+    measureUnit <- GetSourceUnits(srcType)[2]
+
+    # Sorting data
+    if (!is.null(sourceStore[[1]])) {
+      xdata <- sort(unlist(sourceStore), decreasing = TRUE)
+      ydata <- sort(unlist(measStore), decreasing = FALSE)
+    } else {
+      xdata <- NA
+      ydata <- NA
+    }
+
+    figureData <- list(
+      list(
+        x = xdata,
+        y = ydata,
+        mode = "lines+markers",
+        name = "IV curve",
+        line = list("color" = accentColor[[theme]], "width" = 2)
+      )
+    )
+    ivFigure <- list(
+      "data" = figureData,
+      "layout" = list(
+        xaxis = list(
+          "title" = sprintf("Applied %s (%s)", sourceLabel, sourceUnit),
+          "color" = textColor[[theme]],
+          "gridcolor" = gridColor[[theme]]
+        ),
+        yaxis = list(
+          "title" = sprintf("Measured %s (%s)", measureLabel, measureUnit),
+          "color" = textColor[[theme]],
+          "gridcolor" = gridColor[[theme]]
+        ),
+        font = list(color = textColor[[theme]], size = 12),
+        automargin = TRUE,
+        plot_bgcolor = cardColor[[theme]],
+        paper_bgcolor = cardColor[[theme]]
+      )
+    )
+
+    if (!(modeChoice)) { # Single measure case
+      if (measTriggered) {
+
+        # Single data point does not draw graph,
+        # duplicating data to show single data point
```

Comment:
Interesting; why is this step necessary?

Reply:
When the voltage/current value set from the `daqKnob` and `Single Measure` is clicked, we expect having a single data point on the graph like in the python version. However, when there's only 1 data point in `xdata` and `ydata`, the callback was returning an empty graph. Just duplicating values in this scenario solved the issue. However, not sure if it is the best solution.

<img width="1244" alt="Screen Shot 2019-08-08 at 2 55 07 PM" src="https://user-images.githubusercontent.com/20918264/62729745-9a150500-b9ec-11e9-856c-1a9dbffa8b64.png">

Reply:
Ah, OK; thanks for the additional detail. I think we should revisit this sometime soon, but for now it's fine. We can discuss alternatives and then just open a new, tiny PR to update later.